### PR TITLE
Implement Class Tetris. Issue # 24

### PR DESCRIPTION
**Pull request by PashaBeshchuk**

### CODE REVIEW

File: tetris.js [hunk 1]
```diff
@@ -571,6 +571,8 @@ module.exports = {
 				break;
 		}
 		return this.shiftCoordinates(arrayForTetromino, shift);
-
+	},
+	Tetris: class {
```
Comment (zbstof) [inline tetris.js]: please don't start your PRs with such dummy implementations.

File: test/test-createTetromino.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            `)
+        };
+        assert.deepEqual(resultFunction, expected);
+    })
+    it("Create I tetromino", function () {
+        let fieldSize = { x: 5, y: 4 };
+        let resultFunction = tetris.createTetromino(fieldSize);
+        let expected = {
+            typeOfTetromino: "I",
+            rotationPhase: "vertical",
```
Comment (zbstof) [inline test/test-createTetromino.js]: Why vertical?

File: test/test-createTetromino.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+var tetris = require('../tetris');
+describe("Create tetromino", function () {
+    it("Create L tetromino", function () {
+        let fieldSize = { x: 5, y: 2 };
+        let resultFunction = tetris.createTetromino(fieldSize);
+        let expected = {
+            typeOfTetromino: "L",
+            rotationPhase: "right",
+            coordinates: tetris.getCoordinates(`
+                -XXX-
```
Comment (zbstof) [inline test/test-createTetromino.js]: I don't think this matches initial phase

File: test/test-createTetromino.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        assert.deepEqual(resultFunction, expected);
+    })
+    it("Create J tetromino", function () {
+        let fieldSize = { x: 5, y: 2 };
+        let resultFunction = tetris.createTetromino(fieldSize);
+        let expected = {
+            typeOfTetromino: "J",
+            rotationPhase: "left",
+            coordinates: tetris.getCoordinates(`
+                -XXX-
```
Comment (zbstof) [inline test/test-createTetromino.js]: And this

File: test/test-createTetromino.js [hunk 5]
```diff
@@ -94,6 +99,7 @@ describe("Create tetromino", function () {
                 -----
             `)
         };
+        stub.restore()
```
Comment (zbstof) [inline test/test-createTetromino.js]: Why do you need restore here?

Comment (PashaBeshchuk) [reply test/test-createTetromino.js]: If I transfer the line to 104, it will not work

Comment (zbstof) [reply test/test-createTetromino.js]: Why restore only in this test?

File: test/test-createTetromino.js [hunk 6]
```diff
@@ -0,0 +1,105 @@
+var sinon = require('sinon');
+var assert = require('chai').assert;
+var tetris = require('../tetris');
+describe("Create tetromino", function () {
+    let stub = sinon.stub(tetris,"getRandomTypeOfTetromino")
```
Comment (zbstof) [inline test/test-createTetromino.js]: Stub is a bad name for this variable. Don't use such generic names - they are not readable

File: test/test-createTetromino.js (hunk 5 above)
Comment (PashaBeshchuk) [reply test/test-createTetromino.js]: Recommend closing at the end of the test

File: test/test-createTetromino.js (hunk 6 above)
Comment (PashaBeshchuk) [reply test/test-createTetromino.js]: Done

File: test/test-createTetromino.js (hunk 5 above)
Comment (zbstof) [reply test/test-createTetromino.js]: This does not answer my question

Comment (PashaBeshchuk) [reply test/test-createTetromino.js]: Что бы в каждом it не объявлять  let stubForGetRandomTypeOfTetromino = sinon.stub(tetris, "getRandomTypeOfTetromino").
А если в каждом it делать restore, то в следующем it проверка проходить не будет.

**Comment by zbstof** [comment]
Why do you think this test will be invoked last?

PashaBeshchuk <notifications@github.com> 于 2019年7月16日周二 09:56写道：

**Comment by PashaBeshchuk** [comment]
a sequence of tests such

**Comment by PashaBeshchuk** [comment]
Есть вопрос по поводу функции keyboardHandler.
Что бы добраться к функциям moveLeft() или moveRigt() и т.д. нужно в функцию передать класс тетромино или поместить функцию в сам класс.
Как лучше поступить?

**Comment by zbstof** [comment]
Передай всередину object Tetris

вт, 16 лип. 2019, 17:49 користувач PashaBeshchuk <notifications@github.com>
пише:

**Comment by zbstof** [comment]
Add invokation of tick to your moves test. Also tick when tetromino cannot
move down

вт, 16 лип. 2019, 19:50 користувач Стас Бицько <zbstof@gmail.com> пише:

File: test/test-classTetrisMoves.js [hunk 7]
```diff
@@ -44,6 +44,24 @@ describe("All moves tetromino", function () {
...
+        elementTetris.tick()
+        let expectedTick = tetris.getCoordinates(`
+            -------
+            -------
+            --XX---
+            ---X---
+            ---X---
+        `)
+        assert.deepEqual(elementTetris.tetromino.coordinates, expectedTick)
+        elementTetris.tick()
```
Comment (zbstof) [inline test/test-classTetrisMoves.js]: you should also check the field before and after this tick

File: test/test-classTetrisMoves.js [hunk 8]
```diff
@@ -210,6 +282,24 @@ describe("All moves tetromino", function () {
...
+        let expectedTick = tetris.getCoordinates(`
+            -------
+            -------
+            ---X---
+            --XX---
+            --X----
+        `)
+        assert.deepEqual(elementTetris.tetromino.coordinates, expectedTick)
+        elementTetris.tick()
```
Comment (zbstof) [inline test/test-classTetrisMoves.js]: why don't you change the type of tetromino that will appear on the field

File: test/test-classTetrisMoves.js [hunk 9]
```diff
@@ -252,9 +339,26 @@ describe("All moves tetromino", function () {
...
+        let expectedTick = tetris.getCoordinates(`
+            -------
+            --X----
+            --X----
+            --X----
+            --X----
+        `)
+        assert.deepEqual(elementTetris.tetromino.coordinates, expectedTick)
+        elementTetris.tick()
```
Comment (zbstof) [inline test/test-classTetrisMoves.js]: what will happen on the next tick if nothing is pressed?

File: tetris.js [hunk 10]
```diff
@@ -571,6 +571,62 @@ module.exports = {
...
+		}
+		moveDown(){
+			let shift = { x: 0, y: 1 }
+			this.tetromino.moveTetromino(this.field.field, shift)
+		}
+		rotate(){
+			this.tetromino.rotateTetromino(this.field.field)
+		}
+	},
+	keyboardHandler: function (keyCode, tetris){
```
Comment (zbstof) [inline tetris.js]: Lol, no
This will not work. Ask yourself - how this will be called

File: test/test-classTetrisMoves.js (hunk 7 above)
Comment (PashaBeshchuk) [reply test/test-classTetrisMoves.js]: I do a status check before tick
string 39

Comment (zbstof) [reply test/test-classTetrisMoves.js]: you do not check the field currently
check the field

File: test/test-classTetrisMoves.js [hunk 11]
```diff
@@ -410,14 +463,23 @@ describe("All moves tetromino", function () {
...
-            --XX---
-            --XX---
+            ----X--
+            --XXX--
             -------
             -------
             -------
         `)
         assert.deepEqual(elementTetris.tetromino.coordinates, expectedCantMoveDownAndNewTetromino)
+        // let expectedField = tetris.buildField(`
```
Comment (zbstof) [inline test/test-classTetrisMoves.js]: What is that?
Never commit commented code, please

File: test/test-classTetrisMoves.js [hunk 12]
```diff
@@ -0,0 +1,485 @@
...
+        `)
+        assert.deepEqual(elementTetris.tetromino.coordinates, expectedCantMoveDownAndNewTetromino)
+        let expectedField = tetris.buildField(`
+            -------
+            -------
+            ---X---
+            ---X---
+            --XX---
+        `)
+        assert.deepEqual(elementTetris.field.field, expectedField)
```
Comment (zbstof) [inline test/test-classTetrisMoves.js]: and now do another tick

File: test/test-classTetrisMoves.js [hunk 13]
```diff
@@ -0,0 +1,485 @@
+var sinon = require('sinon');
```
Comment (zbstof) [inline test/test-classTetrisMoves.js]: this test is too long. Shorten to 100 lines with all equivalence classes

Comment (PashaBeshchuk) [reply test/test-classTetrisMoves.js]: what exactly cut?

File: test/test-classTetrisMoves.js [hunk 14]
```diff
@@ -218,5 +219,83 @@ describe("All moves tetromino", function () {
...
+        stubForGetRandomTypeOfTetromino.returns("I")
+        elementTetris.tick()
+        let expectedFieldAddOSecond = tetris.buildField(`
+        --------
+        -----X--
+        -----X--
+        ---XXX--
+        --XXXX--
+        XXXX-XXX
+    `)
```
Comment (zbstof) [inline test/test-classTetrisMoves.js]: formatting

File: test/test-classTetrisMoves.js [hunk 15]
```diff
@@ -218,5 +219,83 @@ describe("All moves tetromino", function () {
...
+    elementTetris.tick()
+    let expectedFieldAddOThird = tetris.buildField(`
+        --XXXX--
+        -----X--
+        -----X--
+        ---XXX--
+        --XXXX--
+        XXXX-XXX
+    `)
+    assert.deepEqual(elementTetris.field.field, expectedFieldAddOThird)
```
Comment (zbstof) [inline test/test-classTetrisMoves.js]: do one more tick

File: init.js [hunk 16]
```diff
@@ -0,0 +1,24 @@
+let tetris = require('./tetris');
+module.exports = {
+    init: function () {
```
Comment (zbstof) [inline init.js]: this function does not implement requirements completely

File: test/test-classTetrisMoves.js (hunk 14 above)
Comment (zbstof) [reply test/test-classTetrisMoves.js]: FORMATTING!

File: test/test-classTetrisMoves.js [hunk 17]
```diff
@@ -0,0 +1,301 @@
...
+        assert.deepEqual(elementTetris.field.field, expectedFieldAddOSecond)
+        elementTetris.tick()
+        let expectedFieldAddOThird = tetris.buildField(`
+        --XXXX--
+        -----X--
+        -----X--
+        ---XXX--
+        --XXXX--
+        XXXX-XXX
+    `)
```
Comment (zbstof) [inline test/test-classTetrisMoves.js]: FORMATTING!

File: test/test-classTetrisMoves.js [hunk 18]
```diff
@@ -0,0 +1,301 @@
...
+            --------
+            --------
+            ---XX---
+            ---XX---
+            --------
+        `)
+        assert.deepEqual(elementTetris.tetromino.coordinates, expectedMovesRightTicksO)
+        stubForGetRandomTypeOfTetromino.returns("I")
+        elementTetris.tick()
+        let expectedFieldAddOSecond = tetris.buildField(`
```
Comment (zbstof) [inline test/test-classTetrisMoves.js]: these variable names are meaningless. Check that every variable is actually reflecting what is going on, or you can just reassign single variable over and over in this test

File: test/test-classTetrisMoves.js [hunk 19]
```diff
@@ -0,0 +1,301 @@
+var sinon = require('sinon');
+var assert = require('chai').assert;
+var tetris = require('../tetris');
+describe("All moves tetromino", function () {
+    let stubForGetRandomTypeOfTetromino = sinon.stub(tetris, "getRandomTypeOfTetromino");
+    it("Mini Game", function () {
+        stubForGetRandomTypeOfTetromino.returns("O")
+        let fieldSize = { x: 8, y: 6 }
+        let elementTetris = new tetris.Tetris(fieldSize)
```
Comment (zbstof) [inline test/test-classTetrisMoves.js]: why it's called elementTetris? It's not element in any way.
Just call it game or something

File: test/test-classTetrisMoves.js [hunk 20]
```diff
@@ -0,0 +1,301 @@
...
+        let expectedFieldAddOSecond = tetris.buildField(`
+        --------
+        -----X--
+        -----X--
+        ---XXX--
+        --XXXX--
+        XXXX-XXX
+    `)
+        assert.deepEqual(elementTetris.field.field, expectedFieldAddOSecond)
+        elementTetris.tick()
```
Comment (zbstof) [inline test/test-classTetrisMoves.js]: do one more tick

File: tetris.js (hunk 10 above)
Comment (zbstof) [reply tetris.js]: What is this function doing here? You are declaring the same in init

File: init.js [hunk 21]
```diff
@@ -3,7 +3,7 @@ module.exports = {
     init: function () {
         let fieldSIze = { x: 5, y: 5 };
         let newTetris = new tetris.Tetris(fieldSIze);
-        let tick = setInterval(newTetris.tick(), 1000);
+        let tick = setInterval(newTetris.tick, 1000);
         keyboardHandler: function (keyCode, tetris) {
```
Comment (zbstof) [inline init.js]: keyboard handler is not called anywhere - fix this

File: init.js [hunk 22]
```diff
@@ -0,0 +1,27 @@
+let tetris = require('./tetris');
+module.exports = {
+    init: function () {
+        let fieldSIze = { x: 5, y: 5 };
+        let newTetris = new tetris.Tetris(fieldSIze, this.stopTick());
```
Comment (zbstof) [inline init.js]: you are passing stopTick function incorrectly

File: tetris.js [hunk 23]
```diff
@@ -571,6 +572,49 @@ module.exports = {
...
-
+	},
+	createTetromino: function (fieldSize) {
+		let typeOfTetromino = this.getRandomTypeOfTetromino()
+		let phase = this.startingPhaseOfTetromino(typeOfTetromino)
+		let coodinates = this.initCoordinates(typeOfTetromino, fieldSize)
+		return new this.Tetromino(typeOfTetromino, phase, coodinates)
+	},
+	Tetris: class {
+		constructor(fieldSize, callback) {
```
Comment (zbstof) [inline tetris.js]: callback is too generic a name. Something like gameOverCallback is better

File: init.js [hunk 24]
```diff
@@ -2,9 +2,11 @@ let tetris = require('./tetris');
 module.exports = {
     init: function () {
         let fieldSIze = { x: 5, y: 5 };
-        let newTetris = new tetris.Tetris(fieldSIze, this.stopTick());
+        let newTetris = new tetris.Tetris(fieldSIze, stopTick());
         let tick = setInterval(newTetris.tick, 1000);
-        keyboardHandler: function (keyCode, tetris) {
+
+        keyboardHandler()
```
Comment (zbstof) [inline init.js]: what is this? This signature and line 9 are not consistent

File: init.js [hunk 25]
```diff
@@ -2,9 +2,11 @@ let tetris = require('./tetris');
 module.exports = {
     init: function () {
         let fieldSIze = { x: 5, y: 5 };
-        let newTetris = new tetris.Tetris(fieldSIze, this.stopTick());
+        let newTetris = new tetris.Tetris(fieldSIze, stopTick());
```
Comment (zbstof) [inline init.js]: this is not how you pass a function parameter to a constructor

File: tetris.js [hunk 26]
```diff
@@ -580,25 +580,28 @@ module.exports = {
...
-		constructor(fieldSize, callback) {
+		constructor(fieldSize, gameOverCallback) {
 			this.fieldSize = fieldSize;
 			this.typeOfTetromino = module.exports.getRandomTypeOfTetromino()
 			this.coordinatesOfTetramino = module.exports.initCoordinates(this.typeOfTetromino, this.fieldSize)
 			this.tetromino = module.exports.createTetromino(this.fieldSize)
 			this.field = new module.exports.Field(this.fieldSize)
-			this.callback = callback
+			this.gameOverCallback = gameOverCallback
+			//console.log(this.gameOverCallback)
```
Comment (zbstof) [inline tetris.js]: DO. NOT. COMMIT. COMMENTED CODE.

File: init.js (hunk 22 above)
Comment (PashaBeshchuk) [reply init.js]: I don't know, how I must transfer function stopTick :(
I went through a lot of options.
Via module.exports.init.stopTick
Via module.exports.init.stopTick()
Via init.stopTick()
Via init.stopTick
Via stopTick
Via stopTick()
All varianits is wrong:(

**Comment by zbstof** [comment]
Describe error that you're got for each of the options

**Comment by PashaBeshchuk** [comment]
I want to see code function callback, but he say undefined.
I think, he must init code function.

**Comment by zbstof** [comment]
Same error for all options?

PashaBeshchuk <notifications@github.com> 于 2019年7月26日周五 10:20写道：

File: init.js (hunk 16 above)
Comment (zbstof) [reply init.js]: please read requirements for the function. You did not implement them fully

File: init.js (hunk 25 above)
Comment (zbstof) [reply init.js]: Still not fixed

File: test/test-classTetrisMoves.js (hunk 18 above)
Comment (zbstof) [reply test/test-classTetrisMoves.js]: Still not fixed

File: test/test-keyboardHandler.js [hunk 27]
```diff
@@ -0,0 +1,58 @@
+var sinon = require('sinon');
+var assert = require('chai').assert;
+var tetris = require('../tetris');
+describe("The function that handles the keyboard", function () {
```
Comment (zbstof) [inline test/test-keyboardHandler.js]: remove this file - tetris does not (will not) have such method

File: test/test-classTetrisMoves.js [hunk 28]
```diff
@@ -0,0 +1,312 @@
...
+            --XXXX--
+            -----X--
+            -----X--
+            ---XXX--
+            --XXXX--
+            XXXX-XXX
+        `)
+        assert.deepEqual(gameTetris.field.field, expectedFieldAddOThird)
+        gameTetris.tick()
+        let expectedStopGame = tetris.buildField(`
```
Comment (zbstof) [inline test/test-classTetrisMoves.js]: you are not checking that the game is stopped, you just checking the field.
Please add assert that callback is called

File: test/test-classTetrisMoves.js [hunk 29]
```diff
@@ -0,0 +1,312 @@
+var sinon = require('sinon');
+var assert = require('chai').assert;
+var tetris = require('../tetris');
+var init = require('../init');
+describe("All moves tetromino", function () {
+    let stubForGetRandomTypeOfTetromino = sinon.stub(tetris, "getRandomTypeOfTetromino");
+    it("Mini Game", function () {
+        stubForGetRandomTypeOfTetromino.returns("O")
+        let fieldSize = { x: 8, y: 6 }
+        let gameTetris = new tetris.Tetris(fieldSize)
```
Comment (zbstof) [inline test/test-classTetrisMoves.js]: This constructor expects callback as second argument

File: tetris.js [hunk 30]
```diff
@@ -571,6 +572,49 @@ module.exports = {
...
+		tick() {
+			let shift = { x: 0, y: 1 }
+			if (this.tetromino.canMoveDown(this.field.field)) {
+				this.tetromino.moveTetromino(this.field.field, shift)
+			} else {
+				this.field.addTetrominoToField(this.tetromino.coordinates)
+				this.tetromino = module.exports.createTetromino(this.fieldSize)
+				this.field.cleanFilledRows()
+				if(!module.exports.checkThatTheFieldIsFree(this.tetromino.coordinates, this.field.field)){
+					this.gameOverCallback
```
Comment (zbstof) [inline tetris.js]: what this does, what do you think?

File: init.js [hunk 31]
```diff
@@ -0,0 +1,28 @@
+let tetris = require('./tetris');
+module.exports = {
+    init: function () {
+        let fieldSIze = { x: 5, y: 5 };
+        let newTetris = new tetris.Tetris(fieldSIze, stopTick());
+        let tick = setInterval(newTetris.tick, 1000);
+        keyboardHandler(event.keyCode, newTetris)
```
Comment (zbstof) [inline init.js]: this is not how you call event handler

File: init.js [hunk 32]
```diff
@@ -0,0 +1,28 @@
...
+        function keyboardHandler(keyCode, tetris) {
```
Comment (zbstof) [inline init.js]: incorrect argument list for event handler

File: init.js (hunk 31 above)
Comment (zbstof) [reply init.js]: this is not needed

File: init.js (hunk 32 above)
Comment (zbstof) [reply init.js]: Still incorrect listener singnature
https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#Parameters

File: test/test-classTetrisMoves.js [hunk 33]
```diff
@@ -279,15 +280,15 @@ describe("All moves tetromino", function () {
...
             -----X--
             -----X--
             ---XXX--
             --XXXX--
             XXXX-XXX
         `)
-        assert.deepEqual(gameTetris.field.field, expectedFieldAddOSecond)
+        assert.deepEqual(gameTetris.field.field, expectedFieldAddIThird)
         gameTetris.tick()
         let expectedFieldAddOThird = tetris.buildField(`
```
Comment (zbstof) [inline test/test-classTetrisMoves.js]: you didn't change all other variable names - this has nothing to do with O-tetromino, for example

File: init.js [hunk 34]
```diff
@@ -9,9 +9,8 @@ module.exports = {
         let newTetris = new tetris.Tetris(fieldSIze, stopTick);
         let tick = setInterval(newTetris.tick, 1000);
         window.addEventListener("keydown", keyboardHandler)
-        keyboardHandler(event.keyCode, newTetris)
-        function keyboardHandler(keyCode, tetris) {
-            switch (keyCode) {
+        function keyboardHandler(event, tetris) {
```
Comment (zbstof) [inline init.js]: still not correct

File: init.js [hunk 35]
```diff
@@ -0,0 +1,32 @@
...
+    init: function () {
+        let fieldSIze = { x: 5, y: 5 };
+        let keyCodeRightButton = 39;
+		let keyCodeLeftButton =  37;
+		let keyCodeUpButton = 38;
+		let keyCodeDownButton = 40;
+        let newTetris = new tetris.Tetris(fieldSIze, stopTick);
+        let tick = setInterval(newTetris.tick, 1000);
+        window.addEventListener("keydown", keyboardHandler)
+        function keyboardHandler(event, tetris) {
```
Comment (zbstof) [inline init.js]: not correct signature

**Pull request merged**